### PR TITLE
[mlir][LLVM] Add support for constant struct with multiple fields

### PR DESCRIPTION
**Pull request by Lancern**
Currently `mlir.llvm.constant` of structure types restricts that the structure type effectively represents a complex type -- it must have exactly two fields of the same type and the field type must be either an integer type or a float type.

This PR relaxes this restriction and it allows the structure type to have an arbitrary number of fields.

### CODE REVIEW

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (Dinistro) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:2686]: Shouldn't this check now be recursive, given that one can define nested struct constants?

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (Dinistro) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:2694]:
```suggestion
      if (!isa<IntegerAttr, FloatAttr>(element)) {
```
This namespace is used. Same for other occurrences.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (Dinistro) [inline mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:564]:
```suggestion
   SmallVector<llvm::Constant *, 8> structElements;
```

File: mlir/test/Dialect/LLVMIR/invalid.mlir
Comment (Dinistro) [inline mlir/test/Dialect/LLVMIR/invalid.mlir:1]: Please add a test with a nested struct

**Review by Dinistro — COMMENTED** [review]
Thanks for the change. I dropped a few comments regarding nested structs.

**Review by gysit — COMMENTED** [review]
Can you elaborate on the use case for this change? 

I think the constant operation is only meant to produce constants of non-aggregate types. The complex type is an exception bcs there is some support for builtin complex attributes which produces an array attribute with two elements.

Note that proper struct support is not possible bcs a struct can contain pointers which cannot be modeled as attributes. Additionally, they can be nested, which may be modeled with attributes but things get complex.

Instead if you want to produce a constant struct, the preferred way is to use the undef operation and then set the individual fields of the struct. 

You can test this by writing some example:
```
define { i32, i32 } @test() {
   ret {i32, i32} { i32 -1, i32 0}
}
```
And then call  `./bin/mlir-translate --import-llvm ./test.ll` to import into MLIR:
```
  llvm.func @test() -> !llvm.struct<(i32, i32)> {
    %0 = llvm.mlir.constant(0 : i32) : i32
    %1 = llvm.mlir.constant(-1 : i32) : i32
    %2 = llvm.mlir.undef : !llvm.struct<(i32, i32)>
    %3 = llvm.insertvalue %1, %2[0] : !llvm.struct<(i32, i32)> 
    %4 = llvm.insertvalue %0, %3[1] : !llvm.struct<(i32, i32)> 
    llvm.return %4 : !llvm.struct<(i32, i32)>
  }
```
This represents an struct constant in LLVM dialect. While this particular example could be supported things will fail as soon as there are pointers:
```
  llvm.func @test() -> !llvm.struct<(ptr, i32)> {
    %0 = llvm.mlir.constant(0 : i32) : i32
    %1 = llvm.mlir.zero : !llvm.ptr
    %2 = llvm.mlir.undef : !llvm.struct<(ptr, i32)>
    %3 = llvm.insertvalue %1, %2[0] : !llvm.struct<(ptr, i32)> 
    %4 = llvm.insertvalue %0, %3[1] : !llvm.struct<(ptr, i32)> 
    llvm.return %4 : !llvm.struct<(ptr, i32)>
  }
```
I would thus stick to the current limitations of constant op and represent aggregate constants using a sequence of operations. Otherwise there will be multiple ways of representing the same thing. That being said maybe you have a good argument why you need constant op to support arrays?

**Comment by Lancern** [comment]
@gysit Thanks for your helpful comment on this topic!

> Can you elaborate on the use case for this change?

Well my actual use case is rather minor compared to the proposed changes here. I'm working on [clangir](https://github.com/llvm/clangir) which provides an MLIR dialect for modeling C/C++ language semantics, and recently I'm trying to work out the appropriate LLVMIR for [pointers to member functions (P2MF)](https://en.cppreference.com/w/cpp/language/pointer#Pointers_to_member_functions). The layout of a P2MF is ABI-specific and currently we're focusing on the Itanium ABI which defines the layout as something like `struct { intptr_t, ptrdiff_t }`. The two members are both integers, but their types may differ. I tried directly lower a P2MF constant into `llvm.mlir.constant` but I come across the current limitation on structure constants. That's the full story.

> I think the constant operation is only meant to produce constants of non-aggregate types.

This makes a lot sense and the pointer problem indeed renders a complete support for arbitrary struct impossible. But for the pointer problem, IMHO it is not really related to the changes here -- you can't produce a constant pointer scalar via `llvm.mlir.constant` either, could you? LLVM itself does not have pointer constants (other than null pointer constants) and that's a more related problem IMO. Besides, `llvm.mlir.constant` is already being used for creating aggregates like vectors (but not arrays or structs yet). So one of the motivation of this PR is that I would expect `llvm.mlir.constant` to be able to produce any constant value supported by LLVM.

**Comment by gysit** [comment]
> The layout of a P2MF is ABI-specific and currently we're focusing on the Itanium ABI which defines the layout as something like struct { intptr_t, ptrdiff_t }.  The two members are both integers, but their types may differ.

I would expect that once you are at the LLVM dialect level you probably want to know the size of these integers (e.g. by querying the data layout)? So from an LLVM dialect point of view the problem is about a simple struct with two integer elements?

> This makes a lot sense and the pointer problem indeed renders a complete support for arbitrary struct impossible. But for the pointer problem, IMHO it is not really related to the changes here -- you can't produce a constant pointer scalar via llvm.mlir.constant either, could you? LLVM itself does not have pointer constants (other than null pointer constants) and that's a more related problem IMO. Besides, llvm.mlir.constant is already being used for creating aggregates like vectors (but not arrays or structs yet). So one of the motivation of this PR is that I would expect llvm.mlir.constant to be able to produce any constant value supported by LLVM.

A nullptr is represented by:
`%1 = llvm.mlir.zero : !llvm.ptr`
There are similar operations for undef / poison values (not just pointers).

I am not completely opposed to supporting constant structs. If we do it, we need have a proper verifier and the documentation should clearly specify what kind of structs are supported. E.g. we could initially support structs without pointers and without nesting. 

I just wanted to point out that there are already alternative ways to express struct constants that work out of the box and that support arbitrary structs. The advantage of your PR is mainly that it simplifies generating struct constants and that folding is potentially easier.

**Comment by gysit** [comment]
FYI, there is a related PR in flight https://github.com/llvm/llvm-project/pull/102906.

**Comment by Lancern** [comment]
> So from an LLVM dialect point of view the problem is about a simple struct with two integer elements?

Yes, that's my use case.

> If we do it, we need have a proper verifier and the documentation should clearly specify what kind of structs are supported. E.g. we could initially support structs without pointers and without nesting.

I agree with this. I updated the documentation and listed the constraints on the LLVM struct type of the result constant.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (Lancern) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp]: Updated.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (Lancern) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: Updated.

File: mlir/test/Dialect/LLVMIR/invalid.mlir
Comment (Lancern) [reply mlir/test/Dialect/LLVMIR/invalid.mlir:1]: As suggested by @gysit :

> we could initially support structs without pointers and without nesting.

Should we include support for nested structure constant in this PR?

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (gysit) [inline mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:564]:
```suggestion
    SmallVector<llvm::Constant *> structElements;
```
nit: We usually do not provide a size except there is a specific reason.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (gysit) [inline mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:566]:
```suggestion
    for (auto [elemType, elemAttr] : zip_equal(arrayAttr, structType.getElementTypes()))
```
nit: Would it be possible to use zip_equal here?

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (gysit) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:2728]: Can you add a test for this?

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (gysit) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:2726]:
```suggestion
      Attribute element = arrayAttr[i];
```
nit: here it makes sense to spell the type since it is not on the rhs.

**Review by gysit — COMMENTED** [review]
Thanks for the updates! I have added some comments.

One thing we do not yet check are llvm.struct types nested an llvm.array. I wonder if such a verification could be added without breaking exiting users/tests. If yes, that would be nice to have. However, I believe this is currently not verified either.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (Lancern) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: Updated.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (Lancern) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: Updated.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (Lancern) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp]: Test added.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (Lancern) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp]: Updated.

**Comment by Lancern** [comment]
> One thing we do not yet check are llvm.struct types nested an llvm.array.

Updated the verifier and added a check that ensures every element type within a struct type is either int or float. Also added a test case for this.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (gysit) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:2739]:
```suggestion
      auto elementType = dyn_cast<TypedAttr>(element).getType();
```
Didn't see this before. I think we should dyn_cast here (as it was before). In case the attribute does not implement the TypedAttr interface we want to return an error and not assert.

**Review by gysit — APPROVED** [review]
Nice!

LGTM modulo one last place where I believe a dyn_cast is necessary.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (gysit) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:2739]: True you are right!

Then this PR is ready to go. Do you have commit rights?

Comment (Lancern) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:2739]: > Do you have commit rights?

No.

Comment (gysit) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:2739]: Ok I will land it then.

**Pull request merged**
